### DISCUSSION TITLE
Change the mouse wheel behavior

### DISCUSSION THREAD

**Discussion by openmax** [sourcelair/xterm.js#5063] [Q&A] [answered]
How can i change the escape code sent to a xterm-256color terminal when i use the wheel of mouse ?
now the escape are for up e down key , but i would like to send "page up" e "page down"  
Thank You

**Comment by jerch** [comment]
You could change the sent bytes by altering the encodings here https://github.com/xtermjs/xterm.js/blob/2edc65baaa50f8b99c67f3df6f8d5d93c70a69e6/src/common/services/CoreMouseService.ts#L121

But be aware, that changing standard encodings to something else will render mouse reports unusable for all apps, that expect standard encoding content (which should be 100% of all apps with proper mouse support). Isn't it easier/less error-prone to tell your app to do the "page up/down" scrolling on a proper wheel event?

**Comment by openmax** [comment]
i'm not an expert , but reading -> https://gist.github.com/halcyon/334da650816876d7be4d1bee8a157f25

i come into the Alternate Scroll mode, so  xterm.js use always this mode .. 
can be implemented to return button 4 and 5 too , so i can send the escape command to switch Alternate mode ?

> Wheel mice
>Wheel mice may return buttons 4 and 5.  Those buttons are represented by
>the same event codes as buttons 1 and 2 respectively, except that 64 is
>added to the event code.  Release events for the wheel buttons are not
>reported.  By default, the wheel mouse events are translated to scroll-
>back and scroll-forw actions.  Those actions normally scroll the whole
>window, as if the scrollbar was used.  However if Alternate Scroll mode
>is set, then cursor up/down controls are sent when the terminal is dis-
>playing the alternate screen.  The initial state of Alternate Scroll
>mode is set using the alternateScroll resource.

**Comment by jerch** [comment]
This sounds a bit like an xy issue. At a higher level - what do you want to achieve?

Normally, w'o any mouse reports being requested by app side, the wheel gets not caught by the terminal and will just scroll on the scrollback (if there is any). Thats browser side only.
Once an app requests mouse reports, this mode is disabled and requested mouse events get reported to app side. In this mouse tracking mode app alone is in charge for mouse stuff and appropriate screen state changes from mouse events.

**Comment by openmax** [comment]
The app the i want to use with page up/down is https://github.com/tn5250/tn5250 , 
is a 5250 terminal emulation that use ncurses library ( ncurses intercept mouse movement )  , 
in the log of the app ( https://github.com/tn5250/tn5250/blob/4c76b521a1b466d087d3f334773a251f53ec3d10/curses/cursesterm.c#L1269 ),
when i use the wheel i receive 4 time the escape code for up/down key , 
and in fact the cursor move up/down for 4 line ..  
( 4 time maybe depend on the "fastScrollSensitivity" setting )

is hard for me modify the tn5250 code to convert a 4 time pressed up/down key in a 1 time pressed page up/down ...

**Reply by jerch** [reply]
> when i use the wheel i receive 4 time the escape code for up/down key ,
> and in fact the cursor move up/down for 4 line ..
> ( 4 time maybe depend on the "fastScrollSensitivity" setting )

I am not quite sure atm, where this comes from. Do you see 4 times up/down key sequences on byte level between xterm.js and the PTY? Normally if your app registers for mouse wheel reports, you should only see those requested events in the selected encoding, not any key events instead. Can you check, if your app correctly asks for mouse wheel reports?

**Comment by openmax** [comment]
look better maybe i found my answer.. -> https://tldp.org/HOWTO/NCURSES-Programming-HOWTO/mouse.html

**Reply by jerch** [reply, marked as answer]
Maybe catch here is to correctly ask for mouse reports with `mousemask`?
